lib/handler/compress/gzip.c

```diff
@@ -74,16 +81,16 @@ static size_t compress_chunk(struct st_gzip_context_t *self, const void *src, si
         }
         self->zs.next_out = (void *)(self->bufs.entries[bufindex].base + self->bufs.entries[bufindex].len);
         self->zs.avail_out = (unsigned)(BUF_SIZE - self->bufs.entries[bufindex].len);
-        ret = deflate(&self->zs, flush);
-        assert(ret == Z_OK || ret == Z_STREAM_END);
+        ret = proc(&self->zs, flush);
+        assert(ret == Z_OK || ret == Z_STREAM_END || ret == Z_BUF_ERROR);
```

Comment:
The current implementation of zlib's `inflate` always returns `Z_BUF_ERROR` when `Z_FINISH` is provided and inflating succeed.
The comment in zlib's `inflate.c` describes it as the following:

```
   In this implementation, the flush parameter of inflate() only affects the
   return code (per zlib.h).  inflate() always writes as much as possible to
   strm->next_out, given the space available and the provided input--the effect
   documented in zlib.h of Z_SYNC_FLUSH.  Furthermore, inflate() always defers
   the allocation of and copying into a sliding window until necessary, which
   provides the effect documented in zlib.h for Z_FINISH when the entire input
   stream available.  So the only thing the flush parameter actually does is:
   when flush is set to Z_FINISH, inflate() cannot return Z_OK.  Instead it
   will return Z_BUF_ERROR if it has not reached the end of the stream.
```

At the last of `inflate` function`..
```
    if (((in == 0 && out == 0) || flush == Z_FINISH) && ret == Z_OK)
        ret = Z_BUF_ERROR;
```

So, I believe we should handle `Z_BUF_ERROR` as non-fatal even if there're enough buffer.